Neos.Neos/Classes/Fusion/Cache/AssetChangeHandlerForCacheFlushing.php

```diff
@@ -0,0 +1,98 @@
+            $contentRepository = $this->contentRepositoryRegistry->get(ContentRepositoryId::fromString($contentRepositoryId));
+            foreach ($usages as $usage) {
+                // TODO: Remove when WorkspaceName is part of the AssetUsageProjection
+                $workspaceName = $workspaceNamesByContentStreamId[$contentRepositoryId][$usage->contentStreamId->value] ?? null;
+                if ($workspaceName === null) {
+                    $workspace = $contentRepository->getWorkspaceFinder()->findOneByCurrentContentStreamId($usage->contentStreamId);
+                    if ($workspace === null) {
+                        continue;
+                    }
+                    $workspaceName = $workspace->workspaceName;
+                    $workspaceNamesByContentStreamId[$contentRepositoryId][$usage->contentStreamId->value] = $workspaceName;
+                }
+                //
+
+                $nodeAggregate = $contentRepository->getContentGraph($workspaceName)->findNodeAggregateById($usage->nodeAggregateId);
+                if ($nodeAggregate === null) {
+                    continue;
+                }
+                $flushNodeAggregateRequest = FlushNodeAggregateRequest::create(
+                    $contentRepository->id,
+                    $workspaceName,
+                    $nodeAggregate->nodeAggregateId,
+                    $nodeAggregate->nodeTypeName,
+                    $this->determineParentNodeAggregateIds($contentRepository, $workspaceName, $nodeAggregate->nodeAggregateId, NodeAggregateIds::createEmpty()),
+                );
+
+                $this->contentCacheFlusher->flushNodeAggregate($flushNodeAggregateRequest, CacheFlushingStrategy::ON_SHUTDOWN);
+            }
+        }
+    }
+
+    private function determineParentNodeAggregateIds(ContentRepository $contentRepository, WorkspaceName $workspaceName, NodeAggregateId $childNodeAggregateId, NodeAggregateIds $collectedParentNodeAggregateIds): NodeAggregateIds
+    {
+        $parentNodeAggregates = $contentRepository->getContentGraph($workspaceName)->findParentNodeAggregates($childNodeAggregateId);
+        $parentNodeAggregateIds = NodeAggregateIds::fromArray(
+            array_map(static fn (NodeAggregate $parentNodeAggregate) => $parentNodeAggregate->nodeAggregateId, iterator_to_array($parentNodeAggregates))
+        );
+
+        foreach ($parentNodeAggregateIds as $parentNodeAggregateId) {
+            // Prevent infinite loops
+            if (!$collectedParentNodeAggregateIds->contain($parentNodeAggregateId)) {
+                $collectedParentNodeAggregateIds = $collectedParentNodeAggregateIds->merge(NodeAggregateIds::create($parentNodeAggregateId));
+                $collectedParentNodeAggregateIds = $this->determineParentNodeAggregateIds($contentRepository, $workspaceName, $parentNodeAggregateId, $collectedParentNodeAggregateIds);
+            }
+        }
+
+
+        return $collectedParentNodeAggregateIds;
+    }
+}
```

Comment:
These are the *ancestorNodeAggregateIds*  to be correct.
And, if we touch this again, I would suggest to streamline it a bit. Using iteration instead of recursion would make this easier to read and faster, e.g.:

```php
    private function determineAncestorNodeAggregateIds(ContentRepository $contentRepository, WorkspaceName $workspaceName, NodeAggregateId $childNodeAggregateId): NodeAggregateIds
    {
        $contentGraph = $contentRepository->getContentGraph($workspaceName);
        /** @var array<NodeAggregate> $stack */
        $stack = iterator_to_array($contentGraph->findParentNodeAggregates($childNodeAggregateId));
        $ancestorNodeAggregateIds = [];
        do {
            $nodeAggregate = array_shift($stack);
            $ancestorNodeAggregateIds[] = $nodeAggregate->nodeAggregateId;
            array_push($stack, ...iterator_to_array($contentGraph->findParentNodeAggregates($nodeAggregate->nodeAggregateId)));
        } while ($stack !== []);
        return NodeAggregateIds::fromArray($ancestorNodeAggregateIds);
    }
```

Reply:
btw: Maybe it makes sense to add this method to the `ContentGraphInterface`...

Reply:
We discussed something similar with @kitsunet and @mhsdesign last week. `findAncestorNodes` on ContentGraphInterface. WDYT?

Reply:
Implemented via https://github.com/neos/neos-development-collection/pull/5261